src/Plugin/GraphQL/DataProducer/Taxonomy/TermAutocomplete.php

```diff
@@ -0,0 +1,347 @@
+      return FALSE;
+    }
+    if ($field_storage_config->getSetting('target_type') != 'taxonomy_term') {
+      return FALSE;
+    }
+
+    return TRUE;
+  }
+
+  /**
+   * Gets the vocabularies configured for given field if it is a term field.
+   *
+   * @param string $entity_type
+   *   Entity type of the field to get the vocabularies for.
+   * @param string $bundle
+   *   Bundle of entity type of the field to get the vocabularies for.
+   * @param string $field
+   *   Field name to get the vocabularies for.
+   *
+   * @return string[]
+   *   Vocabularies configured for the field in case it is a term field, null
+   *   otherwise.
+   */
+  protected function getTermFieldVocabularies(string $entity_type, string $bundle, string $field): ?array {
+    // Load field config of given entity type of given bundle. If not obtained,
+    // bail out.
+    if (!$field_config = $this->getFieldConfig($entity_type, $bundle, $field)) {
+      return NULL;
+    }
+
+    // Make sure the field is configured for taxonomy terms.
+    $field_storage_config = $field_config->getFieldStorageDefinition();
+    if (!$this->isTermFieldStorageConfig($field_storage_config)) {
+      return NULL;
+    }
+
+    // Make sure that target vocabularies are configured.
+    $handler_settings = $field_config->getSetting('handler_settings');
+    if (empty($handler_settings['target_bundles'])) {
+      return NULL;
+    }
+
+    // Return list of vocabularies.
+    return $handler_settings['target_bundles'];
+  }
+
+  /**
+   * Gets term items matched against given query for given vocabulary.
+   *
+   * @param string $entity_type
+   *   Entity type the searchable term field is attached to.
+   * @param string $bundle
+   *   Bundle the searchable term field is attached to.
+   * @param string $field
+   *   Field name to search the terms on.
+   * @param string|null $match_string
+   *   String to be matched.
+   * @param bool $prioritize_start_with
+   *   Whether terms which start with matching string should come first.
+   * @param int $limit
+   *   Number of items to be returned.
+   * @param \Drupal\graphql\GraphQL\Execution\FieldContext $context
+   *   The caching context related to the current field.
+   *
+   * @return array
+   *   List of matched terms.
+   */
+  public function resolve(
+    string $entity_type,
+    string $bundle,
+    string $field,
+    ?string $match_string,
+    bool $prioritize_start_with,
+    int $limit,
+    FieldContext $context
+  ): ?array {
+    if ($limit <= 0) {
+      $limit = 10;
+    }
+
+    // Get configured vocabulary. If none is obtained, bail out.
+    if (!$vocabularies = $this->getTermFieldVocabularies($entity_type, $bundle, $field)) {
+      return NULL;
+    }
+
+    // Base of the query selecting term names and synonyms.
+    $query = $this->database->select('taxonomy_term_field_data', 't');
+    $query->fields('t', ['tid']);
+    $query->condition('t.vid', $vocabularies, 'IN');
```

Comment:
we need to add the taxonomy_term_access tag here, right?

and how does that affect caching? Does that mean we cannot cache the result of this resolver?

Reply:
every query is cached based on query and input variables out of the box.
we additionally tag the result by returned term ids and `taxonomy_term_list`

Reply:
I see, so the current user with their access rights would be an input for this query to consider for caching?

How does drupal core implement taxonomy autocompletes and caching?